notifications_utils/clients/otel/utils.py

```diff
@@ -0,0 +1,175 @@
+import time
+from collections.abc import Callable
+from contextlib import AbstractContextManager, contextmanager
+from functools import wraps
+from typing import Any
+
+from opentelemetry.metrics import get_meter
+from opentelemetry.trace import Span, Status, StatusCode, Tracer, get_tracer
+
+default_histogram_bucket = [
+    0.005,
+    0.01,
+    0.025,
+    0.05,
+    0.075,
+    0.1,
+    0.25,
+    0.5,
+    0.75,
+    1.0,
+    2.5,
+    5.0,
+    7.5,
+    10.0,
+    float("inf"),
+]
+
+# Examples of how to use the decorator:
+
+# Example 1: Static attributes (current behavior)
+# @otel_span(attributes={"operation": "send_email"})
+# def send_email(to, subject):
+#     ...
+
+# Example 2: Dynamic attributes based on function arguments
+# @otel_span(attributes=lambda args, kwargs: {"user_id": kwargs.get("user_id")})
+# def process_user(user_id):
+#     ...
+
+# Example 3: Dynamic attributes using both args and kwargs
+# @otel_span(attributes=lambda args, kwargs: {
+#     "first_arg": args[0] if args else None,
+#     "keyword": kwargs.get("keyword")
+# })
+# def do_something(a, keyword=None):
+#     ...
+
+
+def otel_span(
```

Comment:
FWIW Instead of implementing decorators directly, nowadays I tend to use https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager which automagically generates both a decorator and context manager so you can use it as either.

Reply:
Though come to think I'm not sure it would allow you to inspect the args & kwargs of the inner function.

Reply:
But If we don't have any cases where we're actually _using_ the "callable attrs" yet (don't think I've spotted one in the apps), could we, for now, do away with this and instead just use `otel_span_with_status` seeing as it should work as a decorator too. The only remaining difference I can see is the latter expects a `tracer` to be injected and this conjures up its own. Is that really unreconcilable?

Reply:
I'll go with what you want on this front.

I created these functions to try and minimise the repetitive code in api.

The code actually using these functions changed quite a bit over time. I left them with their full functionality, even though not everything currently uses that to try and future proof them a bit.

So up to you really.

Reply:
I've removed otel_span